src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessService.java

```diff
@@ -69,11 +68,12 @@ public LocalDate collect(Mandate mandate, Payment payment) {
         var goCardlessMandateId = mandate.getPaymentProviderMandateId()
                 .map(a -> (GoCardlessMandateId) a)
                 .orElseThrow( () -> new GoCardlessMandateNotConfirmed("mandate id", mandate.getExternalId().toString()));
-        
-        GoCardlessPayment goCardlessPayment = createPayment(payment, goCardlessMandateId);
 
-        persist(goCardlessPayment);
-        return goCardlessPayment.getChargeDate();
+        PaymentProviderPaymentIdAndChargeDate providerIdAndChargeDate = createPayment(payment, goCardlessMandateId);
+        payment.setProviderId(providerIdAndChargeDate.getPaymentProviderPaymentId());
+        payment.setChargeDate(providerIdAndChargeDate.getChargeDate());
```

Comment:
Rather than modifying the existing Payment by `setProviderId` and `setChargeDate`, I would prefer to have a Payment builder which can take a provider id and charge date, as well as take an existing payment to copy the existing fields into a new Payment object. In other words I'd like objects to be immutable where possible. Happy for this to go in a new PR though.

Reply:
Yep, we'll add that to the list of things we'd like to improve upon in the next PR. This is already getting quite big.